torch/csrc/jit/passes/break_continue_transform.cpp

```diff
@@ -0,0 +1,326 @@
+    auto guard_block = new_if->addBlock();
+
+    while (iter != block->nodes().end()) {
+      auto node = *iter++;
+      node->moveBefore(guard_block->return_node());
+    }
+
+    {
+      WithInsertPoint insert(hit_control_flow_block);
+      // NB: insert var scape before transform kind so it is not removed
+      // In a graph like:
+      // for i in range(3):
+      //     if cond == 2:
+      //         if cond == 2:
+      //             m = 2
+      //             break
+      //         k = 1
+      //     else:
+      //         k = 2
+      //     m += k
+      // We transform the inner cond == 2 block to look like:
+      // if cond == 2:
+      //     m = 2
+      //     $did_break = True
+      // else:
+      //     $did_break = False
+      // if $did_break...
+      //    prim::NewVarEscape
+      // else:
+      //    k = 1
+      // For these new if nodes that guard ops after a continue/break may have
+      // occurred, the new variables that are defined need to escape scope.
+      // Otherwise, in the example above, we would error in the m += k call.
+      graph_->insertNode(graph_->create(prim::NewVarEscape, 0));
+      graph_->insertNode(graph_->create(transformKind(), 0));
+    }
+    return handleIf(new_if);
+  }
+
+  void deleteAfterBreakNodes(Block* block, graph_node_list_iterator& iter) {
+    if (iter == block->nodes().end()) {
+      return;
+    }
+    // need to destroy in reverse order so nodes have no uses when destroyed
+    for (auto it = block->nodes().reverse().begin(); it != iter;) {
+      if (*it == block->return_node()) {
+        it++;
+      } else {
+        it.destroyCurrent();
+      }
+    }
+    iter->destroy();
+  }
+
+  void inlineLoopConditionIntoLoopBody(Node* n) {
+    auto body_block = n->blocks().at(0);
+    auto pre_header = n->blocks().at(1);
+    moveBlockBeforeNode(body_block->return_node(), pre_header);
+    body_block->insertOutput(0, pre_header->outputs().at(0));
+    n->eraseBlock(1);
+  }
+
+  void handleLoop(Node* loop_node) {
+    // transform the loop, then ensure that that it does not accidentally
+    // pick up or assign the current transform variable outside of the loop.
+    transformLoop(loop_node);
+    Block* body_block = loop_node->blocks().at(0);
+    graph_->createStore(getVarname(), false_val_)
+        ->insertAfter(body_block->param_node());
+    graph_->createStore(getVarname(), false_val_)
+        ->insertBefore(body_block->return_node());
+  }
+
+  void transformLoop(Node* n) {
+    Block* body_block = n->blocks().at(0);
+    auto ret_status = handleTransforms(body_block);
+
+    // When we're transforming breaks:
+    // the body condition has not yet been inlined. If we we are not breaking
+    // we need to inline the condition block into the end of the loop.
+    // if we might break, we create an if statement and only execute the loop
+    // header if we did not break.
+    // Since we run the continue pass before the break pass,
```

Comment:
A lot of state to keep in mind. How are we going to maintain this pass?

Reply:
Breaks have to either lower the condition as part of the pass, or happen after they are lowered. If they don't do it as part of the pass the graphs will be worse (i.e have an extra if statement), but that's fine. 

I'm not sure how to get around continues happening pre-lowering.

Reply:
Actually i'm not sure if lowering the condition before breaks works